### PR TITLE
Avoid converting to TOSA if the invariants are not respected

### PR DESCRIPTION
**Pull request by josel-amd**
Employ the `SameOperandsAndResultElementType` before the operator is created. Unfortunately, I can employ `verifySameOperandsAndResultElementType` since the operator hasn't been created.

### CODE REVIEW

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 1]
```diff
@@ -513,6 +513,16 @@ func.func @test_pow_f64(%arg0: tensor<13x21x1xf64>, %arg1: tensor<13x21x1xf64>)
 
 // -----
 
+func.func @main_graph(%arg0: tensor<3xf32>, %arg1: tensor<3xi32>) -> (tensor<3xf32>) {
+  // CHECK-LABEL:  func @main_graph
+  // CHECK-SAME:   ([[PARAM_0:%.*]]: tensor<3xf32>, [[PARAM_1:%.*]]: tensor<3xi32>) -> tensor<3xf32> 
+  // CHECK: %{{.*}} = "onnx.Pow"([[PARAM_0]], [[PARAM_1]]) {onnx_node_name = "onnx.Pow_0"} : (tensor<3xf32>, tensor<3xi32>) -> tensor<3xf32>
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:519]:
```suggestion
  // CHECK: "onnx.Pow"([[PARAM_0]], [[PARAM_1]]) {onnx_node_name = "onnx.Pow_0"} : (tensor<3xf32>, tensor<3xi32>) -> tensor<3xf32>
```

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 2]
```diff
@@ -513,6 +513,16 @@ func.func @test_pow_f64(%arg0: tensor<13x21x1xf64>, %arg1: tensor<13x21x1xf64>)
 
 // -----
 
+func.func @main_graph(%arg0: tensor<3xf32>, %arg1: tensor<3xi32>) -> (tensor<3xf32>) {
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:516]: Maybe `test_pow_mixed_types`?

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 3]
```diff
@@ -92,6 +92,19 @@ LogicalResult checkBasicTosaRequirementsForBinaryOps(
 
   Type resultElementType = resultTensorType.getElementType();
 
+  if (op->template hasTrait<
+          ::mlir::OpTrait::SameOperandsAndResultElementType>()) {
+    auto lhsElementType =
+        lhs.getType().template cast<TensorType>().getElementType();
```
[the remark this answers is not in the document]
Comment (josel-amd) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:98]: Right, I moved the code around and forgot to update!

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 2 above)
Comment (josel-amd) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:516]: Fixed!

**Review by TinaAMD — APPROVED** [review]
Looks nice!

**Review by cferry-AMD — APPROVED** [review]
LGTM

**Pull request merged**